### PR TITLE
fix: invalidation of tracker tree DHIS2-14213

### PR DESCRIPTION
**Pull request by teleivo**
## Invalidating Entities

This adds `PersistablesFilter` which collects valid entities (trackedEntities, enrollments, events, relationships) and errors explaining why other entities are invalid. It relies on the validation hooks that run before it doing the bulk of the work. It is only concerned with collecting

* entities which are valid i.e. have no validation errors found by previous hooks
* and entities which have valid parents (on CREATE/UPDATE) or valid children (DELETE)

All entities passing this filter should be persistable (CREATE/UPDATE/DELETE).

Previously, some of this work at least for relationships was done in the `RelationshipsValidationHook`. Its now moved into a single place the `PersistablesFilter`.

## Reporting errors

The persistables filter will return errors for entities which it considers invalid only if its not already clear from the previous validation (hooks).

For example in case

```java
        Setup setup = new Setup()
                .trackedEntity( "xK7H53f4Hc2" ).isInvalid()
                    .enrollment( "t1zaUjKgT3p" ).isInvalid();
```

Since the TEI is invalid due to some validation hook it will already have an error in the report. So its clear to a user why it cannot be deleted. If the TEI were valid 

```java
        Setup setup = new Setup()
                .trackedEntity( "xK7H53f4Hc2" )
                    .enrollment( "t1zaUjKgT3p" ).isInvalid();
```

error `E5001( "trackedEntity" xK7H53f4Hc2 cannot be deleted because "enrollment" t1zaUjKgT3p referenced by it cannot be deleted."`  will be added. This makes it clear why a valid TEI in the payload cannot be deleted.

The same behavior is implemented for CREATE/UPDATE/CREATE_UPDATE where a child can only be created if its parent exists.

2 new generic error codes `E5000` and `E5001` for describing the above error for any entity and its parent/child are added.

### CODE REVIEW

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/DefaultTrackerValidationService.java [hunk 1]
```diff
@@ -108,7 +109,15 @@ private TrackerValidationReport validate( TrackerBundle bundle, List<TrackerVali
             .addErrors( reporter.getErrors() )
             .addWarnings( reporter.getWarnings() );
 
-        removeInvalidObjects( bundle, reporter );
+        PersistablesFilter.Result persistables = filter( bundle, reporter.getInvalidDTOs(),
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/DefaultTrackerValidationService.java:112]: PersistablesFilter has ample javadocs and I also added a package javadoc explaining it. Can you be specific as to where you would like more explanation? 😃

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/PersistablesFilter.java [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+ * {@link TrackerImportStrategy#UPDATE} a valid enrollment can be updated if its
+ * parent the TEI is invalid but exists. Same applies to
+ * {@link TrackerImportStrategy#DELETE} as you cannot delete an entity that does
+ * not yet exist.</li>
+ * <li>An {@link Event} in an event program does not have an
+ * {@link Event#getEnrollment()} (parent) set in the payload. This expects one
+ * to be set during preprocessing. So events in program with or without
+ * registration do not get any special treatment.</li>
+ * <li>{@link TrackerImportStrategy#DELETE} with
+ * {@link org.hisp.dhis.security.Authorities#F_TEI_CASCADE_DELETE} is not
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/PersistablesFilter.java:105]: We do for example check the TEIs enrollments on delete with cascade in https://github.com/dhis2/dhis2-core/blob/cd75cabd8e1d75f1f56788f20f0f80121993e1f0/dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckSecurityOwnershipValidationHook.java#L135

What do you mean by ignore children?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/PersistablesFilter.java [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+ * treated differently than a delete without such authority. Validation should
+ * have already flagged entities in {@code invalidEntities} if they have
+ * children that cannot be deleted by that user.</li>
+ * </ul>
+ */
+class PersistablesFilter
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/PersistablesFilter.java:111]: Maybe the delete and createOrUpdate logic all in one place can be hard to follow in the code.
I found myself to always having to remember what mark was meaning in the context I was following.

Do you think it would make sense to have them separated from the beginning, maybe even having 2 different filters one for the delete and the other for the createOrUpdate?

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/PersistablesFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,821 @@
...
+        // @formatter:off
+        Setup setup = new Setup.Builder()
+                .trackedEntity( "xK7H53f4Hc2" )
+                    .enrollment( "t1zaUjKgT3p" )
+                        .event( "Qck4PQ7TMun" )
+                .trackedEntity( "QxGbKYwChDM" )
+                    .enrollment( "Ok4Fe5moc3N" )
+                .relationship("Te3IC6TpnBB",
+                        trackedEntity("xK7H53f4Hc2"),
+                        event("Qck4PQ7TMun") ).isNotValid()
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/PersistablesFilterTest.java:452]: Just small suggestion, maybe it would be more readable if the `.isNotValid()` goes on the next line, aligned with relationships?

Just suggesting because it is the second time I mistakenly think the event is invalid

**Review by enricocolasante — COMMENTED** [review]
A few questions and some suggestions

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java [hunk 5]
```diff
@@ -188,19 +188,29 @@ public String getUsername()
...
-        return this.enrollments.stream().filter( t -> t.getEnrollment().equals( id ) ).findFirst();
+        return findById( this.enrollments, uid );
+    }
+
+    public Optional<Relationship> getRelationship( String uid )
+    {
+        return findById( this.relationships, uid );
+    }
+
+    private static <T extends TrackerDto> Optional<T> findById( List<T> entities, String uid )
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/TrackerBundle.java:211]: Yes there is, forget about it

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/PersistablesFilterTest.java (hunk 4 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/PersistablesFilterTest.java:452]: Will do 👌

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/ValidationErrorReporter.java [hunk 6]
```diff
@@ -89,7 +90,11 @@ public ValidationErrorReporter( TrackerIdSchemeParams idSchemes, boolean failFas
     {
         this.errors = new ArrayList<>();
         this.warnings = new ArrayList<>();
-        this.invalidDTOs = new EnumMap<>( TrackerType.class );
+        this.invalidDTOs = new EnumMap<>( Map.of(
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/ValidationErrorReporter.java:93]: Since it's an enum map of a collection this initializes the collections. This way you can safely check containment or add to the collection of a type.

**Review by enricocolasante — APPROVED** [review]
Looks good :+1:
